File: guides/micronaut-security-session-database-authentication/java/src/main/java/example/micronaut/repositories/UserJdbcRepository.java
package example.micronaut.repositories;

import example.micronaut.entities.User;
import io.micronaut.core.annotation.NonNull;
import io.micronaut.data.jdbc.annotation.JdbcRepository;
import io.micronaut.data.model.query.builder.sql.Dialect;
import io.micronaut.data.repository.CrudRepository;
import jakarta.validation.constraints.NotBlank;

import java.util.Optional;

@JdbcRepository(dialect = Dialect.POSTGRES) // <1>
public interface UserJdbcRepository extends CrudRepository<User, Long> { // <2>
    Optional<User> findByUsername(@NonNull @NotBlank String username);
}

File: guides/micronaut-security-session-database-authentication/python/src/example/micronaut/repositories/user_jdbc_repository.py
from java.util import Optional
from micronaut.data.jdbc.annotation import JdbcRepository
from micronaut.data.repository import CrudRepository

from ..entities.user import User


@JdbcRepository(dialect="POSTGRES")  # <1>
class UserJdbcRepository(CrudRepository[User, int]):  # <2>
    def findByUsername(self, username: str) -> Optional[User]: ...
